RFC repo.amd.com folder hierarchy (#4414)
ROCm/TheRock: docs/rfcs/RFC0011-Repo-Structure.md

repo.amd.com's open source software release publications need standardization. In scope is the ROCm software ecosystem which spans the ROCm Core SDK, expansions like ROCm-DS, and standalone projects like RVS. Software packaging for this ecosystem needs a well defined hierarchy reflected in the package distribution folder structure. As software is published on repo.amd.com, the planned hierarchy must be extensible by other software ecosystem published by AMD. As a result, this proposal includes the ability to add the AMD GPU driver to this structure in the future.

## Definitions

* Repository Streams
  * nightly - nightly builds from the develop branch
  * prerelease - builds from the release candidate branches
  * stable - GA releases of ROCm with a short term support lifecycle, tagged as ROCm releases.
  * lts - Future long term stability (LTS) releases
* Products
  * Core SDK
  * expansions - SDK built with dependencies on the ROCm Core SDK
  * extras - standalone components part of ROCm
* pyindex - folder name for a central repository for python packages

## Repository Structure

`repo.amd.com` will have the following folder structure:

- **amdgpu** *(reserved for future use)*

> - **amd-repos**

Author: ```suggestion
- **amdrepos**
```

     - packages
       - **Linux Distros [a–z]** 
- **archives** *(unmaintained releasees, for reference only)*
- **rocm** (current rocm folder with non production releases, move to archives in 6 months)
- **rocm-platform**
  - **standard**
    - **nightly** *(Retention policy: 30 dev, 120 nightly)*
      - **pyindex** *(central nightly wheel repository for all ROCm components)*
      - **core**
        - tarball

>         - zip

Author: ```suggestion
        - zip
        - installers
```

        - tarball
        - whl
        - packages
      - **extras-[ROCm-major]** #projects released independently for each ROCm major version
        - Structure options:
          - Flat or per-project folders  
          - Per-project allows S3 bucket permission granularity by group but complicates duplication on `repo.amd.com`
        - Flat structure:
          - All projects share structure
          - Each project has its own S3 bucket
          - All projects must have unique product names (→ unique artifact names)
        - tarball
        - whl
        - packages
      - **pytorch**
        - tarball *(maybe?)*
        - whl
      - **jax**
      - **onnx-runtime**

  - **prerelease** *(Retention policy: 2 years)*
    - Mirrors nightly folder structure
    - Tested by QA
    - Must match structure of `repo.amd.com`

  - **stable**
    - Current ROCm Core release from TheRock
    - **standard** (includes default build packages, asan build packages, default-debug symbol packages, and asan-debug system packages)
    - **rpath** (includes rpath variant of standard packages)

  - **lts**
    - `YYYYMM`
      - Mirrors stable folder structure

## Repository Package

> Allow users to install the ROCm repositories via a convienient package. The package provides

Author: ```suggestion
Allow users to install the all ROCm repositories via a convienient package. The package provides
```

## Definitions

- Repository Streams
  - nightly - nightly builds from the develop branch
  - prerelease - builds from the release candidate branches
  - stable - GA releases of ROCm with a short term support lifecycle, tagged as ROCm releases.
  - lts - Future long term stability (LTS) releases
- Products
  - Core SDK
  - expansions - SDK built with dependencies on the ROCm Core SDK
  - extras - standalone components part of ROCm
- pyindex - folder name for a central repository for python packages

## Repository Structure

`repo.amd.com` will have the following folder structure:

- **amdgpu** *(reserved for future use)*
- **amdrepos**
  - packages
    - **Linux Distros [a–z]**
- **archives** *(unmaintained releasees, for reference only)*
- **rocm** (current rocm folder with non production releases, move to archives in 6 months)

> - **rocm-platform**

Reviewer: With the proposed alternative, we could go with `rocm` instead of `rocm-platform`.

Author: outdated

  - **standard**

    - **nightly** *(Retention policy: 30 dev, 120 nightly)*

>       - **pyindex** *(central nightly wheel repository for all ROCm components)*

Reviewer: At least for ROCm we aim for more than a single index, see #5289. While I assume `pyindex` should be based on the new multi-arch packages, we need to clarify if the aggregated index should be based on multi-arch index or the backward compatible multi-arch index.

Author: Basically, one multi arch pyindex and separate standalone pyindex for each llvm architecture target?

Reviewer: The above issue proposes something different:

* multi-arch index, needs an device extra to be specified manually (e.g. `pip install rocm[device-gfx942]`)
* multi-arch index, `pip install rocm` pulling in all device extras

We didn't considered per-family indexes.

Author: Updated — pyindex now contains two siblings: multi-arch/ (explicit device extras) and multi-arch-compat/ (backward-compatible, pip install rocm pulls in all extras). New "Python Indices" section spells out the contract and references #5289.

      - **core**
        - tarball
        - zip
        - installers

>         - whl

Reviewer: Same as above, this needs to be aligned with #5289.

Author: outdated

        - packages
          - **Linux Distros [a–z]**
      - **windows**
        - MSI and EXE files for Windows
      - **expansions [a–z]**
        - tarball
        - whl
        - packages
      - **extras-[ROCm-major]** #projects released independently for each ROCm major version
        - Structure options:
          - Flat or per-project folders
          - Per-project allows S3 bucket permission granularity by group but complicates duplication on `repo.amd.com`
        - Flat structure:
          - All projects share structure
          - Each project has its own S3 bucket
          - All projects must have unique product names (→ unique artifact names)
        - tarball
        - whl
        - packages
      - **pytorch**
        - tarball *(maybe?)*
        - whl
      - **jax**
      - **onnx-runtime**

  - **prerelease** *(Retention policy: 2 years)*

    - Mirrors nightly folder structure
    - Tested by QA

>     - Must match structure of `repo.amd.com`

Reviewer: ```suggestion
    - Must match structure of nightly packages
```
?

Author: outdated

  - **stable**

    - Current ROCm Core release from TheRock
    - **standard** (includes default build packages, asan build packages, default-debug symbol packages, and asan-debug system packages)

>     - **rpath** (includes rpath variant of standard packages)

Reviewer: Seems we already have `standard` at the top level. With that we would have e.g. `rocm-platform/standard/stable/standard/core/packages/ubuntu22/`  where the first standard is the top-level namespace and the second is the build variant. @saadrahim, please confirm if this is intentional or if one should be dropped.

Author: This was intentional for rpath and regular path packages. I am not sure if we have rpath packages yet.

Reviewer: My question here is if we really want `stable`/{..}/stable` or if we should de-duplicate `stable` in the path.

Author: When I view it, I don't see a duplicate stable folder. I am going to check with AI to parse it again.

Author: You were right, I corrected the duplication. removed the top level standard

>   - **lts**

Reviewer 2: might as well spell out 'long term support' here

Author: addressed

---
author: Saad Rahim (saadrahim)
created: 2026-04-08
modified: 2026-05-28
status: draft
---

# ROCm software ecosystem package repository structure

## Related RFCs

- **RFC0008** — multi-arch Python wheels and device extras (source of the

>   `pyindex/multi-arch/` model used here).

Reviewer: ```suggestion
  `whl{,next/` model used here).
```
See above comments proposing the naming change.

  > stream taxonomy for extras — it only states that extras may publish
  > "nightly / pre-release builds … to a staging repository for early
  > validation" alongside ordinary releases. The canonical streams on
  > `repo.amd.com` are the ones defined in this RFC
  > (`nightly`/`rc`/`stable`, with `ltsrc`/`lts` reserved).
  > Per-extra publishing maps into those streams as follows: nightly /
  > staging builds → `nightly/`, pre-release builds intended for QA →
  > `rc/`, GA releases → `stable/`. Each extra's release notes
  > record which stream a given build landed in.
- Products
  - Core SDK
  - expansions - SDK built with dependencies on the ROCm Core SDK
    - HPC SDK - a ROCm expansion released on its own cadence, pinned to a single ROCm version per release
  - extras - standalone components part of ROCm
- pyindex - top-level folder for the ROCm Python package indices. Contains
  two sibling indices (see "Python Indices" section):
  - `pyindex/multi-arch/` — multi-arch index requiring explicit device
    extras (e.g. `pip install rocm[device-gfx942]`)
  - `pyindex/multi-arch-compat/` — backward-compatible multi-arch index

>     where `pip install rocm` pulls in all device extras automatically

Reviewer: Discussed this with @ScottTodd and we suggest to follow PyTorch (https://download.pytorch.org/whl/) closer and name it `whl` and `whl-next`:

```suggestion
- `whl/` - backward-compatible multi-arch ROCm Python packages index
  where `pip install rocm` pulls in all device extras automatically
- `whl-next/` - multi-arch ROCm Python packages index requiring explicit device
  extras (e.g. `pip install rocm[device-gfx942]`)
```

Skipping the top level has the advantage of shorter URLs which we would very much prefer.

## Stream Subdomains

Each release stream is hosted at its own subdomain of `repo.amd.com`,
using the pattern `<stream>.repo.amd.com`. The subdomain *is* the
stream selector — stream-scoped paths sit at the root of the subdomain
rather than under a `<stream>/` prefix on the parent domain.

| Stream    | Subdomain               | Status            |
| :-------- | :---------------------- | :---------------- |
| nightly   | `nightly.repo.amd.com`  | required at v1    |
| rc        | `rc.repo.amd.com`       | required at v1    |
| stable    | `stable.repo.amd.com`   | required at v1    |
| ltsrc     | `ltsrc.repo.amd.com`    | future (reserved) |
| lts       | `lts.repo.amd.com`      | future (reserved) |

> | archives  | `archives.repo.amd.com` | required at v1    |

Reviewer: ```suggestion
| nightly   | `nightly.repo.amd.com`  | required at v2    |
| rc        | `rc.repo.amd.com`       | required at v1    |
| stable    | `stable.repo.amd.com`   | required at v1    |
| ltsrc     | `ltsrc.repo.amd.com`    | future (reserved) |
| lts       | `lts.repo.amd.com`      | future (reserved) |
| archives  | `archives.repo.amd.com` | required at v2    |
```
I think we should prioritize stable and rc over the others.

> `rocm/` folder under each stream subdomain (see next section) share
> the same name but live on different hosts (`repo.amd.com/rocm/` vs
> `<stream>.repo.amd.com/rocm/`) and are unrelated. The bare-domain
> `rocm/` is the legacy folder slated for archive; the per-subdomain
> `rocm/` is the new ROCm platform tree.

### Structure on each `<stream>.repo.amd.com` subdomain

The folder tree below is **identical across every stream subdomain**
(`nightly`, `rc`, `stable`, `ltsrc`, `lts`). Stream-specific
specializations (per-stream content variants, retention) are captured
in *Per-stream specializations* further down.

- **amdgpu/** *(reserved for future use; same tree, future GPU driver
  artifacts)*
- **rocm/**
  - **pyindex/** — **central** PEP 503 simple index for the entire
    stream. A single `pyindex/` serves wheels from every wheel-
    producing area in the stream (`core/`, `expansions/`,
    wheel-producing extras, `pytorch/`, `jax/`, `onnx-runtime/`).
    There is no per-package `pyindex/`. Required sub-folders:
    - **one/** — single-arch variant (user picks device extras)
    - **all/** — all-arch variant (`pip install rocm` pulls in every

>       device extra automatically)

Reviewer: ```suggestion
  - **whl/** - **central** backward-compatible ROCm Python packages index for the entire
    stream. A single index serves wheels from every wheel-
    producing area in the stream (`core/`, `expansions/`,
    wheel-producing extras, `pytorch/`, `jax/`, `onnx-runtime/`).
  - **whl-next/** - **central** ROCm Python packages index for the entire
    stream. A single index serves wheels from every wheel-
    producing area in the stream (`core/`, `expansions/`,
    wheel-producing extras, `pytorch/`, `jax/`, `onnx-runtime/`).   
```

As proposed above. I further consider to create a PEP 691 compatible index instead of 503, so I would leave it open.

    See Python Indices section.
  - **core/**
    - tarball
    - zip
    - installers
    - **whl** — must publish two variants: (1) single-arch wheels
      where the user picks device extras at install time, and (2)
      all-arch wheels where `pip install rocm` pulls in every device
      extra automatically. Internal folder layout, filenames, and
      sub-paths are implementation details left to the publish

>       tooling — not consumed by humans.

Reviewer: ```suggestion
    - **whl**
    - **whl-next**
```

    - packages
      - **Linux Distros [a–z]**
  - **windows/** — MSI and EXE files for Windows.
  - **expansions [a–z]/** *(e.g. **hpc-sdk** — see HPC SDK Release Model section)*
    - tarball
    - **whl** — same two-variant rule as `core/whl` (single-arch +

>       all-arch). Internal layout is implementation-defined.

Reviewer: ```suggestion
    - **whl**
    - **whl-next**
```

  ```

> **Naming note:** the `one/` and `all/` sub-folders under `pyindex/`
> are required and stable — they are the entry points that
> `amdrocm-repo` templates into `--index-url` and are part of the
> public contract. The layout *under* each (sharding, filename
> conventions) and the layout of the underlying `whl/` folders that
> back them remain implementation details left to the publish tooling
> and are not required to be human-readable.

Both variants ship under every stream that publishes wheels
(`nightly`, `rc`, `stable`; not `ltsrc`/`lts` until LTS exists), and
both are built from the same underlying wheel set — `pyindex/all/`
simply republishes the entry-point wheels (`rocm`, `torch`,
`torchvision`, …) with `device-all` added as an automatic requirement,
plus links to the unmodified device wheels in `pyindex/one/` so
storage is not duplicated.

**Wheel-only ROCm dependency rule (applies to every wheel published on
`repo.amd.com`):** any package distributed as a Python wheel — Core
SDK wheels, expansion wheels, extras wheels, third-party AI fork
wheels — **must obtain its ROCm dependency exclusively through

> `pyindex/`**. That is:

Reviewer: ```suggestion
`whl{,-next}/`**. That is:
```

- Wheel `install_requires` / `Requires-Dist` entries that resolve a
  ROCm dependency must resolve to **other ROCm wheels served by

>   `pyindex/one/` or `pyindex/all/`**.

Reviewer: ```suggestion
  `whl/` or `whl-next/`**.
```

- Wheels must not depend on, assume the presence of, or trigger the
  installation of ROCm via any **non-wheel** channel: rpm/deb native
  packages, tarballs, runfile installers, container base images,
  out-of-band scripts, etc. If a wheel needs HIP, a ROCm library, or
  any other ROCm component, that component must itself be available as
  a wheel through `pyindex/`.

> - A `pip install <pkg> --index-url https://<stream>.repo.amd.com/pyindex/{one,all}/`

Reviewer: ```suggestion
- A `pip install <pkg> --index-url https://<stream>.repo.amd.com/whl{,-next}/`
```

  must complete the entire ROCm install chain. The user must not be
  required to run `yum`, `apt`, a `.run` installer, or extract a
  tarball as a prerequisite.
- Native package installs (rpm/deb/tar/runfile) remain the supported
  path for users who want ROCm from native packages — but the **wheel
  path is fully self-contained** and never crosses over into the
  native-package world.

This rule keeps every `pip`-driven install fully resolvable from a
single `--index-url`, makes the wheel ecosystem reproducible across
distros (and inside container images that have no system package

> manager), and ensures the central `pyindex/` is the single source of

Reviewer: ```suggestion
manager), and ensures the central index is the single source of
```

truth for the wheel side of the ROCm ecosystem.

Future direction: WheelNext (`uv pip install` with a wheel-variant
provider backed by `rocm-bootstrap`) is the long-term plan and will

> eventually make `pyindex/all/` unnecessary. Until that lands and is

Reviewer: ```suggestion
eventually make `whl/` unnecessary. Until that lands and is
```

widely adopted, both variants must coexist.

**Native multi-arch (parallel mechanism):** `pyindex/one/` and

> `pyindex/all/` cover the **wheel** side of multi-arch. The **native-package**

Reviewer: This is a bit confusing to me, maybe replace with
```suggestion
**Native multi-arch (parallel mechanism):** `whl/` and
`whl-next/` cover the Python side of ROCm. The **native-package**
```
?

side is covered by `rocm-kpack` per RFC0008 — host and device code are
split into separate packages, with device code shipped either as
per-architecture packages (`amdrocm-<library>-gfx<arch>`) or loaded at
runtime from kpack archives. These device packages and the
architecture-family meta-packages that group them (`rocm-gfx94X`,
`rocm-gfx90X`, etc.) are published under
`rocm/<stream>/core/packages/<distro>/` alongside the host
ROCm Core SDK packages — they are not a separate folder. Wheel

> multi-arch (via the two `pyindex` variants) and native multi-arch (via

Reviewer: ```suggestion
multi-arch (via the two index variants) and native multi-arch (via
```

RFC0008 owns the full multi-arch contract for both sides.

## Third Party AI Forks

`pytorch`, `jax`, and `onnx-runtime` are **ROCm forks/builds of upstream
third-party AI frameworks**, not first-party AMD projects. They are
published on `repo.amd.com` so users can pick up ROCm-enabled wheels
without having to build them locally, but the upstream project owns the
source of truth and the release cadence for the framework itself.

Rules that apply to all third-party AI forks:

- **Upstream tracking:** each entry mirrors a specific upstream release
  (or upstream nightly), with ROCm patches applied on top. Metadata in
  every artifact records the upstream version and the ROCm version it
  was built against.
- **Streams:** published only under `nightly/`, `rc/`, and
  `stable/`. Not published under `ltsrc/` or `lts/` — long-term-support
  guarantees do not extend to third-party fork builds.
- **Artifact format:** `whl` only. No tarballs, no native distro
  packages — users install via `pip` from the matching ROCm wheel index.
- **Dependency rule:** framework wheels must depend **only on Python
  wheels of the ROCm Core SDK** (published under `core/whl/` and

>   surfaced through the central `pyindex/one/` and `pyindex/all/`).

Reviewer: ```suggestion
  surfaced through the central `whl/` and `whl-next/`).
```

## Definitions

*
* nightly - nightly builds from the develop branch
* prerelease - builds from the release candidate branches
* stable - GA releases of ROCm with a short term support lifecycle, tagged as ROCm releases.
* lts - Future long term stability (LTS) releases
* expansions - SDK built with dependencies on the ROCm Core SDK
* pyindex - folder name for a central repository for python packages

## Repository Structure

`repo.amd.com` will have the following folder structure:

- **amdgpu** *(reserved for future use)*  
- **archives** *(unmaintained releasees, for reference only)*
- **rocm** (current rocm folder with non production releases, move to archives in 6 months)
- **rocm-ecosystem**
  - **nightly** *(Retention policy: 30 dev, 120 nightly — builds.amd.com)*
    - **pyindex** *(central nightly wheel repository for all ROCm components)*
    - **core**

>       - tarball

Author: ```suggestion
      - archive
```

Author: ```suggestion
      - tarball
      - zip
      
```

---
author: Saad Rahim (saadrahim)
created: 2026-04-08
modified: 2026-04-08
status: draft
---

# ROCm software ecosystem package repository structure

## Overview

repo.amd.com's open source software release publications need standardization. In scope is the ROCm software ecosystem which spans the ROCm Core SDK, expansions like ROCm-DS, and standalone projects like RVS. Software packaging for this ecosystem needs a well defined hierarchy reflected in the package distribution folder structure. As software is published on repo.amd.com, the planned hierarchy must be extensible by other software ecosystem published by AMD. As a result, this proposal includes the ability to add the AMD GPU driver to this structure in the future.

## Definitions

> *

Author: ```suggestion
```

* nightly - nightly builds from the develop branch
* prerelease - builds from the release candidate branches
* stable - GA releases of ROCm with a short term support lifecycle, tagged as ROCm releases.
* lts - Future long term stability (LTS) releases
* expansions - SDK built with dependencies on the ROCm Core SDK
* pyindex - folder name for a central repository for python packages

## Repository Structure

`repo.amd.com` will have the following folder structure:

- **amdgpu** *(reserved for future use)*  
- **archives** *(unmaintained releasees, for reference only)*
- **rocm** (current rocm folder with non production releases, move to archives in 6 months)
- **rocm-ecosystem**

>   - **nightly** *(Retention policy: 30 dev, 120 nightly — builds.amd.com)*

Author: ```suggestion
  - **nightly** *(Retention policy: 120 nightly — builds.amd.com)*
```

    - **pyindex** *(central nightly wheel repository for all ROCm components)*
    - **core**
      - tarball
      - zip

>       

Author: ```suggestion
```

      - whl
      - packages
    - **Linux Distros [a–z]**
    - **windows**
      - MSI and EXE files for Windows
    - **expansions [a–z]**
      - tarball
      - whl

>       - packages

Author: Use same structure as core.

    - **TBD** *(extras or individual releases)*
      - Structure options:

>         - Flat or per-project folders  

Author: ```suggestion
        - Flat folder structure i.e. projectA.lib shares share with projectBlib.
```

Author: Requirement, we must not have naming conflicting. If naming conflicts, should be caught in CI.

Author: Implementation comment: /rocm /devpytorch can be mapped into /pyindex

Author: per component repositoritories and folder structure

Need to create a rocm-repo.rpm and rocm-repo.deb to create appropriate files in /etc/yum.repos.d/ or /etc/apt/

A central authority needs to maintain this package. It will likely need to be available in a GitHub repo.

Author: repo meta data needs a separate job if we want a flat folder structure. Note this was a design decision not to do this.

        - Per-project allows S3 bucket permission granularity by group but complicates duplication on `repo.amd.com`
      - Flat structure:
        - All projects share structure
        - Each project has its own S3 bucket
        - All projects must have unique product names (→ unique artifact names)
      - tarball
      - whl
      - packages
    - **pytorch**
      - tarball *(maybe?)*
      - whl
    - **jax**
    - **onnx-runtime**

  - **prerelease** *(Retention policy: 2 years — builds.amd.com)*
    - Mirrors nightly folder structure
    - Tested by QA
    - Must match structure of `repo.amd.com`

  - **stable** *(repo.amd.com)* or **sts** *(short-term support)*
    - Current ROCm Core release from TheRock

>   - **lts** *(repo.amd.com)*

Author: Does not imply immediate action or design. Just to reserve the name for future use.

>   - **lts** *(repo.amd.com)*

Author: ```suggestion
  - **lts**
```

## Definitions

* nightly - nightly builds from the develop branch
* prerelease - builds from the release candidate branches
* stable - GA releases of ROCm with a short term support lifecycle, tagged as ROCm releases.
* lts - Future long term stability (LTS) releases
* expansions - SDK built with dependencies on the ROCm Core SDK
* pyindex - folder name for a central repository for python packages

## Repository Structure

`repo.amd.com` will have the following folder structure:

- **amdgpu** *(reserved for future use)*  
- **archives** *(unmaintained releasees, for reference only)*
- **rocm** (current rocm folder with non production releases, move to archives in 6 months)

> - **rocm-ecosystem**

Author: ```suggestion
- **rocm-platform**
```
